subprojects/docs/src/docs/userguide/dep-man/01-core-dependency-management/dependency_verification.adoc

```diff
@@ -0,0 +1,849 @@
+
+The `trusted-key` element works similarly to the <<sec:trusting-artifacts,trusted-artifact>> element:
+
+- `group`, the group of the artifact to trust
+- `name`, the name of the artifact to trust
+- `version`, the version of the artifact to trust
+- `file`, the name of the artifact _file_ to trust
+- `regex`, a boolean saying if the `group`, `name`, `version` and `file` attributes need to be interpreted as regular expressions (defaults to `false`)
+
+[WARNING]
+====
+You should be careful when trusting a key globally: try to limit it to the appropriate groups or artifacts:
+
+- a valid key may have been used to sign artifact `A` which you trust
+- later on, the key is stolen and used to sign artifact `B`
+
+It means you can trust the key `A` for the first artifact, probably only up to the released version before the key was stolen, but not for `B`.
+====
+
+[[sec:ignoring-keys]]
+=== Specifying key servers and ignoring keys
+
+Gradle will automatically download the public keys required to verify a signature.
+For this it uses a list of well known and trusted key servers (the list may change between Gradle versions, please refer to the implementation to figure out what servers are used by default).
+
+You can explicitly set the list of key servers that you want to use by adding them to the configuration:
+
+[source,xml]
+----
+<?xml version="1.0" encoding="UTF-8"?>
+<verification-metadata>
+   <configuration>
+      <verify-metadata>true</verify-metadata>
+      <verify-signatures>true</verify-signatures>
+      <key-servers>
+         <key-server uri="hkp://my-key-server.org"/>
+         <key-server uri="https://my-other-key-server.org"/>
+      </key-servers>
+   </configuration>
+</verification-metadata>
+----
+
+Despite this, it's possible that a key is not available:
+
+- because it wasn't published to a public key server
+- because it was lost
+
+In this case, you can ignore a key in the configuration block:
+
+[source,xml]
+----
+<?xml version="1.0" encoding="UTF-8"?>
+<verification-metadata>
+   <configuration>
+      <verify-metadata>true</verify-metadata>
+      <verify-signatures>true</verify-signatures>
+      <ignored-keys>
+         <ignored-key id="abcdef1234567890" reason="Key is not available in any key server"/>
+      </ignored-keys>
+   </configuration>
+</verification-metadata>
+----
+
+As soon as a key is ignored, it will not be used for verification, even if the signature file mentions it.
+However, if the signature cannot be verified with at least one other key, Gradle will mandate that you provide a checksum.
+
+[[sec:local-keyring]]
+=== Exporting keys for faster verification
+
+Gradle automatically downloads the required keys but this operation can be quite slow and requires everyone to download the keys.
+To avoid this, Gradle offers the ability to use a local keyring file containing the required public keys.
+
+If the `gradle/verification-keyring.gpg` file is present, Gradle will search for keys there in priority.
```

Comment:
Single-file keyring seems to be a bad design (e.g. https://dev.gnupg.org/T295 ), because Gradle would have to load that file on each execution, and each modification would have to re-write the file completely (or something like that).

Caching the file in a daemon does not sound trivial as one might launch build with different Java versions which would result in multiple daemons for the same project directory.

@oehme , what do you think?

PS. Does Gradle cache the downloaded public keys in a project-independent folder?

Reply:
Ok so we have different things in play here:

- Gradle caches downloaded keys globally, cross version
- whenever we see a key we need to download
  - at the beginning of the build, we load the keyring file: there's an _in-memory- cache for those keys
  - if the key is present in the keyring, we use it.
  - if it's not in the key file, we download it and put it in the cross-version cache

This file is just an accelerator based on actual dogfooding. It doesn't make things slower like the link you mentioned (which is just about GPG), it makes things faster and avoids everybody spamming key servers (which isn't nice).

Reply:
I see. This makes sense.